contributors/design-proposals/storage/csi-migration.md

```diff
@@ -661,6 +695,93 @@ with minimal disruption to users would be:
+```
+func ProbePluginsWithCSIMigrationDisabled(plugins []volume.VolumePlugin) []volume.VolumePlugin
+func ProbeAttachablePluginsWithCSIMigrationDisabled(plugins []volume.VolumePlugin) []volume.VolumePlugin
+```
+They will [1] check plugin specific migration feature flags as well as [2] the overall
+CSIMigrationInTreeOff feature flag to determine whether ProbeVolumePlugins
+of a legacy in-tree plugin will get invoked. Kubelet's ProbeVolumePlugins and Kube
+Controller Manager's ProbeAttachableVolumePlugins, ProbeExpandableVolumePlugins and
+ProbeControllerVolumePlugins will call the above instead of directly probing in-tree plugins
+that will be migrated.
+
+Once CSIMigrationInTreeOff feature flag and the plugin specific migration
+flags get enabled by default, the build tag `nolegacyproviders` can be enabled on
+pkg/volume/util/csimigration/plugins.go so that no in-tree plugins get probed.
+
+### Detection of migration status for a plugin
+
+Code paths that need to check migration status of a plugin, for example,
+in provisionClaimOperationExternal, findDeletablePlugin, etc. will need to
+depend on a common utility function:
+```
+func IsCSIMigrationEnabledForPluginByName(pluginName string) bool
+```
+that will be introduced in pkg/volume/csi_migration.go
+
+Note that a per-plugin member function of the form IsMigratedToCSI cannot be used
+since [1] a plugin object for in-tree plugins will typically be nil and [2] the
+code implementing IsMigratedToCSI will be removed as part of disabling plugin code.
+
+### Handling of nil Plugin Object returned by FindPluginBySpec/FindPluginByName
+
+Once an in-tree plugin is no longer probed, all the VolumePluginMgr
+functions of the form Find*PluginBySpec/Find*PluginByName will return a nil plugin
+object to code paths that invoke those functions. For example, invocation of:
+
+1. FindProvisionablePluginByName in the pv controller will return nil for a migrated
+plugin that is no longer probed in ProbeControllerVolumePlugins.
+2. FindExpandablePluginBySpec in the expand controller will return nil for a migrated
+plugin that is no longer probed in ProbeExpandableVolumePlugins.
+3. FindAttachablePluginBySpec in the attach/detach controller will return nil for
+a migrated plugin that is no longer probed in ProbeAttachableVolumePlugins.
+
+Each such code path will need to be audited and modified such that:
+1. Code preceding the invocation of Find*Plugin routines performs the necessary
+translation from in-tree to CSI so that when a Find*Plugin routine is invoked,
+it is passed a translated CSI PV rather than an in-tree PV.
+2. Upon receiving a nil plugin, the code checks migration status of a plugin based
+on spec or name rather than throwing an error.
+
+### Enhancements in Controllerss handling Persistent Volumes
+
+#### AttachDetach Controller
+When CSIMigrationInTreeOff is enabled, translation of volume specs from in-tree to
```

Comment:
is this behavior different from when `CSIMigrationInTreeOff` is off? We should always do translation at the same place.

IIUC the only behavior this new flag should actually "toggle" is "probing the in-tree plugin". All the other changes you are planning to make were to support that working; but the flag shouldn't be toggling any other behavior.

Reply:
We will have at least one additional set of areas where the behavior will be different when `CSIMigrationInTreeOff` is disabled but `CSIMigration` is enabled: all the functions that invoke `nodeUsingCSIPlugin` need to pass legacy/untranslated PVs in the generated functions.

So I was thinking of not performing the translations "early" as described for `CSIMigrationInTreeOff` when `CSIMigrationInTreeOff` is disabled. If we perform the translations early, we may not have enough context to translate back to in-tree PVs around the current invocations of `nodeUsingCSIPlugin` as we will not know if we are dealing with a "native" CSI PV or a "translated" in-tree => CSI PV. 

We can depend on a heuristic to translate back to in-tree: if [i] `nodeUsingCSIPlugin` returns false [ii] `CSIMigration` is enabled [iii] `CSIMigrationInTreeOff` is disabled and [iv] csi driver is a migrated driver, we can always translate CSI PV to in-tree. However if someone was expecting to see an error in this scenario when starting with a "native" CSI PV, they won't see the error which may be confusing.

Overall, both approach feels a bit convoluted. Thoughts?

Reply:
`nodeUsingCSIPlugin` should trivially return true when `CSIMigrationInTreeOff=true`. Since this is basically a commit, the node should not have a different "migration status" than the controller

Reply:
A third/better alternative: we can mark the PVs that are translated from in-tree to CSI "early" with an annotation and use that to translate back to in-tree when `nodeUsingCSIPlugin` returns true during Phase 1 (`CSIMigrationInTreeOff=false` and `CSIMigration=true`)